src/backend/distributed/planner/local_distributed_join_planner.c

```diff
@@ -0,0 +1,490 @@
+							  ConversionCandidates *conversionCandidates,
+							  PlannerRestrictionContext *plannerRestrictionContext)
+{
+	RangeTableEntryDetails *localRTECandidate = NULL;
+	RangeTableEntryDetails *distributedRTECandidate = NULL;
+
+	if (list_length(conversionCandidates->localTableList) > 0)
+	{
+		localRTECandidate = linitial(conversionCandidates->localTableList);
+	}
+	if (list_length(conversionCandidates->distributedTableList) > 0)
+	{
+		distributedRTECandidate = linitial(conversionCandidates->distributedTableList);
+	}
+
+	if (LocalTableJoinPolicy == LOCAL_JOIN_POLICY_PREFER_LOCAL)
+	{
+		return localRTECandidate ? localRTECandidate : distributedRTECandidate;
+	}
+	else if (LocalTableJoinPolicy == LOCAL_JOIN_POLICY_PREFER_DISTRIBUTED)
+	{
+		return distributedRTECandidate ? distributedRTECandidate : localRTECandidate;
+	}
+	else
+	{
+		bool allRangeTableEntriesHaveUniqueIndex = AllRangeTableEntriesHaveUniqueIndex(
+			conversionCandidates->distributedTableList);
+
+		if (allRangeTableEntriesHaveUniqueIndex)
+		{
+			return distributedRTECandidate ? distributedRTECandidate : localRTECandidate;
+		}
+		else
+		{
+			return localRTECandidate ? localRTECandidate : distributedRTECandidate;
+		}
+	}
+}
+
+
+/*
+ * AllRangeTableEntriesHaveUniqueIndex returns true if all of the RTE's in the given
+ * list have a unique index.
+ */
+static bool
+AllRangeTableEntriesHaveUniqueIndex(List *rangeTableEntryDetailsList)
+{
+	RangeTableEntryDetails *rangeTableEntryDetails = NULL;
+	foreach_ptr(rangeTableEntryDetails, rangeTableEntryDetailsList)
+	{
+		if (!rangeTableEntryDetails->hasConstantFilterOnUniqueColumn)
+		{
+			return false;
+		}
+	}
+	return true;
+}
+
+
+/*
+ * RemoveFromConversionCandidates removes an element from
+ * the relevant list based on the relation id.
+ */
+static void
+RemoveFromConversionCandidates(ConversionCandidates *conversionCandidates, Oid relationId)
+{
+	if (IsRelationLocalTableOrMatView(relationId))
+	{
+		conversionCandidates->localTableList =
+			list_delete_first(conversionCandidates->localTableList);
+	}
+	else
+	{
+		conversionCandidates->distributedTableList =
+			list_delete_first(conversionCandidates->distributedTableList);
+	}
+}
+
+
+/*
+ * ShouldConvertLocalTableJoinsToSubqueries returns true if we should
+ * convert local-dist table joins to subqueries.
+ */
+static bool
+ShouldConvertLocalTableJoinsToSubqueries(Query *query, List *rangeTableList,
+										 Oid resultRelationId,
+										 PlannerRestrictionContext *
+										 plannerRestrictionContext)
+{
+	if (LocalTableJoinPolicy == LOCAL_JOIN_POLICY_NEVER)
+	{
+		/* user doesn't want Citus to enable local table joins */
+		return false;
+	}
+	if (!ContainsTableToBeConvertedToSubquery(rangeTableList, resultRelationId))
+	{
+		return false;
+	}
+
+	plannerRestrictionContext = FilterPlannerRestrictionForQuery(
+		plannerRestrictionContext, query);
+	if (IsRouterPlannable(query, plannerRestrictionContext))
+	{
+		return false;
```

Comment:
Would be good to have a `DEBUG1` message or such stating that joins with local tables is skipped because it is router plannable